Use Tailwind & Emotion for more components w/ Babel transform & macro.

DFurnes opened the pull request:
### What's this PR do?

This pull request adds [Emotion's Babel transform](https://emotion.sh/docs/@emotion/babel-preset-css-prop) so that we don't need to remember to manually add the `@jsx` pragma/import to files when we use it. I've also found [this very cool macro](https://github.com/ben-rogerson/twin.macro) that makes it dead-simple to use Tailwind classes as part of an Emotion `css` prop or component!

I've also tried using the "Styled Components" API (for [Emotion](https://emotion.sh/docs/styled) or the Tailwind `tw` macro) in this PR, and I'm really digging it!! It lets you easily make "visual" primitives within a component, like this:

```js
const ArticleTitle = tw.h1`text-black font-league-gothic font-normal text-2xl md:text-4xl lg:text-5xl lg:text-5xl text-center uppercase`;
const ArticleSubtitle = tw.p`text-black font-source-sans font-bold md:text-xl text-center`;
const ArticleLede = styled.div`
  ::after {
    content: '';
    ${tw`block my-12 h-1`}
    background: linear-gradient(…);
  }
`;

const ArticleHeader = ({ title, subtitle, children }) => (
  <ArticleLede>
    <ArticleTitle>{title}</ArticleTitle>
    {subtitle ? <ArticleSubtitle>{subtitle}</ArticleSubtitle> : null}
    {children ? <div className="mx-3 my-6 py-3">{children}</div> : null}
  </ArticleLede>
);
```

I've used these tools to update `ArticleHeader` (extracted from a general page styles & markup scattered in a few different places), `Embed`, `Card`, and `PlaceholderText` to be styled with Tailwind and/or Emotion (rather than our old SCSS includes).

### How should this be reviewed?

I'd recommend reviewing this commit-by-commit:

🔌 Adds [`@emotion/babel-preset-css-prop`](https://emotion.sh/docs/@emotion/babel-preset-css-prop). 625987f

💅 Adds [`twin.macro`](https://github.com/ben-rogerson/twin.macro) for Tailwind + Emotion synergy! 0ae43a8

📄 Extract `ArticleHeader` from general page markup, using Styled Components. 25399c8

🖼 Updates `Embed` to use Tailwind/Emotion for styling. 0aec7b2

💳 Updates `Card` to use Tailwind for styling. 97f9d8c

⏳ Updates `PlaceholderText` to use Emotion for styling. 5d799e9

Then just some house-keeping in b2e3417 and 8d5a21e.

### Any background context you want to provide?

> This is nice and all, but why now?!

Next.js limits us to [a single global CSS entry point](https://err.sh/next.js/css-global), and so we'd need to update all of our (S)CSS includes to either Emotion or Tailwind styles (already the plan!) or [CSS Modules](https://github.com/css-modules/css-modules). I'll have more on this in my server-side rendering RFC later this quarter!

### Relevant tickets

References [Pivotal #171650956](https://www.pivotaltracker.com/story/show/171650956).

### Checklist

- [ ] This PR has been added to the relevant Pivotal card.
- [ ] Documentation added for new features/changed endpoints.
- [ ] Added screenshots of front-end changes on small, medium, and large screens.
- [ ] Added appropriate feature/unit tests.

mendelB commented on resources/assets/components/utilities/ArticleHeader.js:
```diff
@@ -0,0 +1,45 @@
+  }
+`;
+
+const ArticleTitle = tw.h1`text-black font-league-gothic font-normal text-2xl md:text-4xl lg:text-5xl lg:text-5xl text-center uppercase`;
```
*`lg:text-5xl` is defined twice here (which is catastrophic).

DFurnes replied:
UNACCEPTABLE!!! 🍋

weerd replied:
I think we should aim to standardize using an alphabetical system cause I've seen this before, definitely done it before and will happen again in the future. Doesn't break anything, but would be ideal to minimize especially when it turns into a long line of classes... easier to manage if in alphabetical order.

The general rule I've followed is to use alphabetical order, ignoring pseudo states and media query prefixes (media queries organized in increasing device order), so the underlying classname is what is considered. (I think I've done classes with numbers in them at the end so `text-center` before `text-4xl` or whatever. So for this example:

```
font-league-gothic font-normal text-black text-center text-2xl md:text-4xl lg:text-5xl lg:text-5xl uppercase
```

DFurnes replied:
Sounds reasonable. I guess I'm going to become very good at alphabetization. 😨

DFurnes replied:
Removed the duplicate class in 7de37c8 & alphabetized in 24b6ae7.

mendelB commented on resources/assets/components/utilities/ArticleHeader.js:
```diff
@@ -0,0 +1,45 @@
+      #fcd116 100%
+    );
+  }
+`;
```
Pretty darn cool!

weerd commented:
> I've also found this very cool macro that makes it dead-simple to use Tailwind classes as part of an Emotion css prop or component!

That's pretty awesome 🎉

mendelB commented on resources/assets/components/pages/GeneralPage/general-page.scss:
```diff
@@ -9,65 +9,13 @@
 
   .general-page__author-bios {
     border-top: 2px solid #dcdcdc;
-    margin: theme('spacing.6') theme('spacing.3') 0;
```
What happened to this margin? (and the margin in the `general-page__authors`?)

DFurnes replied:
I moved this onto the children container, rather than an explicit top-margin on this element. This felt a little cleaner to think about where those margins are getting added.

#### Before:

<img width="622" alt="Screen Shot 2020-03-13 at 10 34 23 AM" src="https://user-images.githubusercontent.com/583202/76630396-52240680-6516-11ea-8022-02a3e91a54fa.png">


#### After:

<img width="599" alt="Screen Shot 2020-03-13 at 10 33 46 AM" src="https://user-images.githubusercontent.com/583202/76630411-55b78d80-6516-11ea-9b60-9755d17e37ad.png">

DFurnes replied:
And here's an example with multiple authors & this new styling:

<img width="621" alt="Screen Shot 2020-03-13 at 10 41 14 AM" src="https://user-images.githubusercontent.com/583202/76630945-3a994d80-6517-11ea-8d07-b9d5c46f3b00.png">

weerd replied:
Does it look ok on small screens?

DFurnes replied:
Yup:

<img width="318" alt="Screen Shot 2020-03-13 at 10 50 43 AM" src="https://user-images.githubusercontent.com/583202/76631745-91535700-6518-11ea-86e1-24e9367da8ac.png">

weerd replied:
👌

weerd commented:
> I've also tried using the "Styled Components" API (for Emotion or the Tailwind tw macro) in this PR, and I'm really digging it!! It lets you easily make "visual" primitives within a component, like this:

I'm a bit less convinced on this, but just may need to try it. It feels a bit like it obscures the markup too much, a bit disjointing and could be confusing since the primitives looks like separate components 🤔

mendelB commented on resources/assets/components/utilities/Card/Card.js:
```diff
@@ -2,19 +2,26 @@ import React from 'react';
 import PropTypes from 'prop-types';
 import classnames from 'classnames';
 
-import './card.scss';
```
I noticed we didn't replace the `.affiliate-content` styling. I don't believe we actively add those anymore, but I don't think we've formally deprecated this yet?

DFurnes replied:
Oh, I really thought we had. 🙈 That partnership has ended right? I'll double-check.

DFurnes replied:
Alright, we got the green light to remove these! Removed in 489840bcbefb4d22b55270f5905269a4c8a3e961:

#### Before:

<img width="692" alt="Screen Shot 2020-03-13 at 10 54 01 AM" src="https://user-images.githubusercontent.com/583202/76636444-6d474400-651f-11ea-9a82-a06e4d621bd9.png">

#### After:

<img width="694" alt="Screen Shot 2020-03-13 at 11 39 06 AM" src="https://user-images.githubusercontent.com/583202/76636402-5bfe3780-651f-11ea-9f67-d90ffbf6c53b.png">

mendelB commented:
> Next.js limits us to a single global CSS entry point, and so we'd need to update all of our (S)CSS includes to either Emotion or Tailwind styles (already the plan!) or CSS Modules

I know you mentioned having more on this soon, but would you mind giving a quick, dumbed-down explanation of what this means exactly for the Mendel's in the crowd? 😁 🙌

DFurnes commented on package.json:
```diff
@@ -125,6 +131,7 @@
     "redux-logger": "^3.0.6",
     "redux-thunk": "^2.3.0",
     "sixpack-client": "^2.1.0",
+    "twin.macro": "^1.0.0-alpha.7",
```
😂

DFurnes commented:
> > I've also tried using the "Styled Components" API (for Emotion or the Tailwind tw macro) in this PR, and I'm really digging it!! It lets you easily make "visual" primitives within a component, like this:
> 
> I'm a bit less convinced on this, but just may need to try it. It feels a bit like it obscures the markup too much, a bit disjointing and could be confusing since the primitives looks like separate components 🤔

Totally fair! I've went back and forth on 'em in the past. My argument for them is that (like the semantic class names of yore), it lets us label pieces of a component in a way that's easier to skim and perhaps re-use (without the "ceremony" of exporting to a completely standalone component).

Let's mull over it for a bit, but definitely not something we have to use.

Here's a side-by-side of how this would look with plain Tailwind/Emotion:

```js
const ArticleTitle = tw.h1`text-black font-league-gothic font-normal text-2xl md:text-4xl lg:text-5xl lg:text-5xl text-center uppercase`;
const ArticleSubtitle = tw.p`text-black font-source-sans font-bold md:text-xl text-center`;
const ArticleLede = styled.div`
  ::after {
    content: '';
    ${tw`block my-12 h-1`}
    background: linear-gradient(…);
  }
`;

const ArticleHeader = ({ title, subtitle, children }) => (
  <ArticleLede>
    <ArticleTitle>{title}</ArticleTitle>
    {subtitle ? <ArticleSubtitle>{subtitle}</ArticleSubtitle> : null}
    {children ? <div className="mx-3 my-6 py-3">{children}</div> : null}
  </ArticleLede>
);
```

```js
const styles = css`
  ::after {
    content: '';

    ${tw`block h-1 my-12`}
    background: linear-gradient(
      to right,
      #23b7fb 0%,
      #23b7fb 33.3%,
      #4e2b63 33.3%,
      #4e2b63 66.6%,
      #fcd116 66.6%,
      #fcd116 100%
    );
  }
`;

const ArticleHeader = ({ title, subtitle, children }) => (
  <div css={styles}>
    <h1 classNames="font-league-gothic font-normal text-black text-center text-2xl md:text-4xl lg:text-5xl uppercase">
      {title}
    </h1>
    {subtitle ? (
      <p classNames="font-source-sans font-bold text-black text-center md:text-xl">
        {subtitle}
      </p>
    ) : null}
    {children ? <div className="mx-3 my-6 text-center">{children}</div> : null}
  </div>
);
```

DFurnes commented:
> I know you mentioned having more on this soon, but would you mind giving a quick, dumbed-down explanation of what this means exactly for the Mendel's in the crowd? 😁 🙌

They add some really cool development (hot-reloading) & production (optimization) features that don't work well when you're importing multiple `.css` files that end up in the "global" namespace, and so they restrict you to a single CSS import (from your JS).

For example, if we happened to write a `.card` rule in multiple files and hot-reloaded one, it'd swap ordering in the DOM and perhaps affect style precedence so the page might look different than it would on initial load.

Luckily, we're already heading in this direction (with Tailwind and Emotion) and so it's not a huge imposition to change – just means we'd need to double-down on finishing that work before we could server-render a particular component.

mendelB commented:
> > I know you mentioned having more on this soon, but would you mind giving a quick, dumbed-down explanation of what this means exactly for the Mendel's in the crowd? 😁 🙌
> 
> They add some really cool development (hot-reloading) & production (optimization) features that don't work well when you're importing multiple `.css` files that end up in the "global" namespace, and so they restrict you to a single CSS import (from your JS).
> 
> For example, if we happened to write a `.card` rule in multiple files and hot-reloaded one, it'd swap ordering in the DOM and perhaps affect style precedence so the page might look different than it would on initial load.
> 
> Luckily, we're already heading in this direction (with Tailwind and Emotion) and so it's not a huge imposition to change – just means we'd need to double-down on finishing that work before we could server-render a particular component.

Ah neat. Thank you -- this is super helpful! And looking forward to the RFC!

DFurnes commented:
Added in a quick fix to a bug we introduced when changing the campaign page background to white in #1905, where the gallery cards would sort-of melt into the page ([discussed here](https://dosomething.slack.com/archives/CJ340DEEL/p1584115145022900)):

#### Before:

<img width="994" alt="Screen Shot 2020-03-13 at 12 08 18 PM" src="https://user-images.githubusercontent.com/583202/76638902-9ec20e80-6523-11ea-85c2-b434fa793760.png">

#### After:

<img width="998" alt="Screen Shot 2020-03-13 at 12 07 45 PM" src="https://user-images.githubusercontent.com/583202/76638916-a386c280-6523-11ea-89de-86ca31a84232.png">

weerd commented:
Awesome. Thanks for the notes in your [comment](https://github.com/DoSomething/phoenix-next/pull/1965#issuecomment-598787876).

I think I still lean towards the plain Tailwind/Emotion, but willing to experiment!

I just wonder, if part of the reason of using Tailwind was to also get away from the decision making of semantic class naming of yore, this seems like it's bringing us back to that to a certain degree, lol! And we'll be discussing and debating in PRs what primitive names make the most sense 😅
